core/src/syscall/precompiles/blake2b/compress/mix.rs

```diff
@@ -0,0 +1,281 @@
+            let d = (d_hi as u64) << 32 | d_lo as u64;
+            let d = d.rotate_right(32);
+            d_lo = d as u32;
+            d_hi = (d >> 32) as u32;
+
+            // c = c + d.
+            (c_lo, c_hi) = self.c_plus_d.populate(record, c_lo, c_hi, d_lo, d_hi);
+
+            // b = (b ^ c).rotate_right(24).
+            b_lo = self.b_xor_c_lo.populate(record, b_lo, c_lo);
+            b_hi = self.b_xor_c_hi.populate(record, b_hi, c_hi);
+            let b = (b_hi as u64) << 32 | b_lo as u64;
+            let b = b.rotate_right(24);
+            b_lo = b as u32;
+            b_hi = (b >> 32) as u32;
+        }
+
+        // Second 4 steps.
+        {
+            // a = a + b + y.
+            (a_lo, a_hi) = self.a_plus_b_2.populate(record, a_lo, a_hi, b_lo, b_hi);
+            (a_lo, a_hi) = self
+                .a_plus_b_2_add_y
+                .populate(record, a_lo, a_hi, y_lo, y_hi);
+
+            // d = (d ^ a).rotate_right(16).
+            d_lo = self.d_xor_a_2_lo.populate(record, d_lo, a_lo);
+            d_hi = self.d_xor_a_2_hi.populate(record, d_hi, a_hi);
+            let d = (d_hi as u64) << 32 | d_lo as u64;
+            let d = d.rotate_right(16);
+            d_lo = d as u32;
+            d_hi = (d >> 32) as u32;
+
+            // c = c + d.
+            (c_lo, c_hi) = self.c_plus_d_2.populate(record, c_lo, c_hi, d_lo, d_hi);
+
+            // b = (b ^ c).rotate_right(63).
+            b_lo = self.b_xor_c_2_lo.populate(record, b_lo, c_lo);
+            b_hi = self.b_xor_c_2_hi.populate(record, b_hi, c_hi);
+            (b_lo, b_hi) = self
+                .b_xor_c_2_rotate_right_63
+                .populate(record, b_lo, b_hi, 63);
+        }
+
+        let result = [a_lo, a_hi, b_lo, b_hi, c_lo, c_hi, d_lo, d_hi];
+        let result_u64 = [
+            (result[1] as u64) << 32 | result[0] as u64,
+            (result[3] as u64) << 32 | result[2] as u64,
+            (result[5] as u64) << 32 | result[4] as u64,
+            (result[7] as u64) << 32 | result[6] as u64,
+        ];
+
+        assert_eq!(result_u64, mix(input_u64));
+        self.result = result.map(Word::from);
+        result
+    }
+
+    pub fn eval<AB: SP1AirBuilder>(
+        builder: &mut AB,
+        input: [Word<AB::Var>; MIX_INPUT_SIZE * 2],
+        cols: MixOperation<AB::Var>,
+        is_real: AB::Var,
+    ) {
+        builder.assert_bool(is_real);
+        let mut a_lo = input[0];
+        let mut a_hi = input[1];
+        let mut b_lo = input[2];
+        let mut b_hi = input[3];
+        let mut c_lo = input[4];
+        let mut c_hi = input[5];
+        let mut d_lo = input[6];
+        let mut d_hi = input[7];
+        let x_lo = input[8];
+        let x_hi = input[9];
+        let y_lo = input[10];
+        let y_hi = input[11];
+
+        // First 4 steps.
+        {
+            // a = a + b + x.
+            AddOperationU64::<AB::F>::eval(builder, a_lo, a_hi, b_lo, b_hi, cols.a_plus_b, is_real);
```

Comment:
I think your life would be easier if there was an `U64` type similar to `Word` that basically had 8 AB::Var elements instead of 4. Then you wouldn't have to have all this hi/lo stuff everywhere and could just keep passing around the `U64`

Reply:
I've created a new struct `WordU64` for this.